### PR TITLE
Add enumerateChildrenByName, walk node tree with callback

### PR DESCRIPTION
**Pull request by heliclei**

### CODE REVIEW

File: cocos/2d/CCNode.h [hunk 1]
```diff
@@ -789,18 +789,26 @@ class CC_DLL Node : public Ref
...
      *
      * @param name  A  string that indentifies the node.
      */
     virtual void setName(const std::string& name);
     /**
+     * Search the children to perform processing for nodes which share a name.
+     * @param name  the name to search for
+     * @param callback, A callback to execute on nodes that match the name parameter. The callback takes the following arguments:
+     *  node: A node that matches the name.
+     *  stop: A pointer to a Boolean variable. Your callback can set this to YES to terminate the enumeration.
```
Comment (dumganhar) [inline cocos/2d/CCNode.h]: `YES` -> `true`

File: cocos/2d/CCNode.h [hunk 2]
```diff
@@ -789,18 +789,26 @@ class CC_DLL Node : public Ref
...
+     */
+     virtual void enumChildNodesByName(const std::string& name, std::function<void(Node* node, bool* stop)> callback);
```
Comment (dumganhar) [inline cocos/2d/CCNode.h]: Please use `const &` for `callback` parameter.

File: samples/cpp-tests/Classes/NodeTest/NodeTest.cpp [hunk 3]
```diff
@@ -354,6 +355,70 @@ void Test6::addAndRemove(float dt)
...
+    addChild(sp11);
+    addChild(sp12);
+    addChild(sp20);
+    addChild(sp21);
+    addChild(sp22);
+
+    auto fn1 = std::bind(&NodeEnumChildByNameTest::runAction1, this, std::placeholders::_1, std::placeholders::_2);
+    auto fn2 = std::bind(&NodeEnumChildByNameTest::runAction2, this, std::placeholders::_1, std::placeholders::_2);
+    
+    enumChildNodesByName("sister1", fn1);
```
Comment (dumganhar) [inline samples/cpp-tests/Classes/NodeTest/NodeTest.cpp]: How about using `lambda` for this test? That will simplify our codes.

Comment (heliclei) [inline samples/cpp-tests/Classes/NodeTest/NodeTest.cpp]: good idea~

File: tests/cpp-tests/Classes/NodeTest/NodeTest.cpp [hunk 4]
```diff
@@ -354,6 +355,65 @@ void Test6::addAndRemove(float dt)
...
+
+    addChild(sp10);
+    addChild(sp11);
+    addChild(sp12);
+    addChild(sp20);
+    addChild(sp21);
+    addChild(sp22);
+
+    enumChildNodesByName("sister1", 
+        [&](Node* node, bool* stop)
```
Comment (dumganhar) [inline tests/cpp-tests/Classes/NodeTest/NodeTest.cpp]: I think `&` isn't needed.

Comment (heliclei) [inline tests/cpp-tests/Classes/NodeTest/NodeTest.cpp]: correct~

**Comment by heliclei** [comment]
[ci rebuild]

File: cocos/2d/CCNode.cpp [hunk 5]
```diff
@@ -733,6 +732,25 @@ Node* Node::getChildByName(const std::string& name)
     return nullptr;
 }
 
+void Node::enumChildNodesByName(const std::string& name, const std::function<void(Node* node, bool* stop)>& callback)
```
Comment (minggo) [inline cocos/2d/CCNode.cpp]: Node::enumChildNodesByName -> Node::emulateChild ?

Comment (minggo) [inline cocos/2d/CCNode.cpp]: Node::emulateChildren?

**Comment by ricardoquesada** [comment]
guys, this is a great feature.
But was it needed for v3.0 ?
I really do thing that names are better that tags, but adding names now it means that we need to deprecate tags... since names replace tags.

So, we need to deprecate the `addChild(child, z, tag)` API... deprecate `setTag()`, deprecate `getTag()`....

**Comment by darkdukey** [comment]
Maybe use tag for searching a group of node?

**Comment by ricardoquesada** [comment]
mmmm... confusing. why both ? which is better?  users are going to ask those questions and more.

we need to keep the API simple:
- simple to understand
- simple to maintain

From the Zen of python:
http://legacy.python.org/dev/peps/pep-0020/

```
There should be one-- and preferably only one --obvious way to do it.
```

**Comment by heliclei** [comment]
Thanks, Riq.

**Comment by heliclei** [comment]
I think deprecate tag is a good idea.

**Comment by dumganhar** [comment]
Yep, since we don't have enough time for the coming release, we will remove this feature.
We need to think more about the pros and cons of `tag` and `name` in the version after the coming release. :)

**Comment by dumganhar** [comment]
@ricardoquesada , sorry for the merge. :)

**Comment by heliclei** [comment]
OK, I will revert this PR.

**Comment by ricardoquesada** [comment]
@dumganhar no problem
This is going to be nice feature.... and we want to promote it once it is ready. So it makes sense to postpone, and add it v3.1. 

TODO:
- We need more tests (including performance tests)
- update all our samples.
- Update actions APi (?)
- what else ??? Does the new API conver all the search patterns supported by SpriteKit ?

**Pull request merged**